### PR TITLE
Allow server-only in server targets  and client-only in client components targets to be available

### LINKED ISSUE #43700: [Next13 App dir] Bug with server-only being used in an API route handler
### Verify canary release

- [X] I verified that the issue exists in the latest Next.js canary release

### Provide environment information

    Operating System:
      Platform: darwin
      Arch: arm64
      Version: Darwin Kernel Version 22.0.0: Mon Aug  1 06:32:20 PDT 2022; root:xnu-8792.0.207.0.6~26/RELEASE_ARM64_T6000
    Binaries:
      Node: 16.15.0
      npm: 8.5.5
      Yarn: N/A
      pnpm: N/A
    Relevant packages:
      next: 13.0.6
      eslint-config-next: 13.0.6
      react: 18.2.0
      react-dom: 18.2.0

### Which area(s) of Next.js are affected? (leave empty if unsure)

App directory (appDir: true)

### Link to reproduction - Issues with a link to complete (but minimal) reproduction code will be addressed faster

https://github.com/KolbySisk/next-server-only-bug

### To Reproduce

1. Create a client component that makes a request to an api route (A form that posts data to an api route, for example)
2.`npm i server-only`
3. Add `import "server-only";` to the api route.
4. Run and make the request to the api route.

See error:

`You're importing a component that needs server-only. That only works in a Server Component but one of its parents is marked with "use client", so it's a Client Component.`

### Describe the Bug

API routes are handled only on the server, so I would expect `import "server-only";` to work fine.

More realistically, I have a function which includes sensitive information that I want to ensure only runs on the server. It can be used in both a server component, and an API route, given the context.

I'm also curious how Next even knows about the relationship between a client component and an API route handler. 

### Expected Behavior

Code is executed as expected without the error.

### Which browser are you using? (if relevant)

_No response_

### How are you deploying your application? (if relevant)

_No response_

<sub>[NEXT-1385](https://linear.app/vercel/issue/NEXT-1385/[next13-app-dir]-bug-with-server-only-being-used-in-an-api-route)</sub>

### LINKED ISSUE #52833: import "server-only" and "use server" throws when used in middleware
### Verify canary release

- [X] I verified that the issue exists in the latest Next.js canary release

### Provide environment information

```bash
Operating System:
      Platform: linux
      Arch: x64
      Version: #22 SMP Tue Jan 10 18:39:00 UTC 2023
    Binaries:
      Node: 16.17.0
      npm: 8.15.0
      Yarn: 1.22.19
      pnpm: 7.1.0
    Relevant packages:
      next: 13.1.7-canary.10
      eslint-config-next: N/A
      react: 18.2.0
      react-dom: 18.2.0
```

### Which area(s) of Next.js are affected? (leave empty if unsure)

Middleware / Edge (API routes, runtime)

### Link to the code that reproduces this issue or a replay of the bug

https://codesandbox.io/p/sandbox/beautiful-montalcini-vdnznm?file=%2Fmiddleware.ts%3A2%2C1

### To Reproduce

Just open the reproduction example.

### Describe the Bug

1. Can't import "server-only" files into middleware.

2. Importing "use server" files into middleware handles it as client file. So for example environment isn't available and code that relies on it breaks.

### Expected Behavior

The explained problems not happening.

### Which browser are you using? (if relevant)

_No response_

### How are you deploying your application? (if relevant)

_No response_

<sub>[NEXT-1616](https://linear.app/vercel/issue/NEXT-1616/import-server-only-and-use-server-throws-when-used-in-middleware)</sub>

### LINKED ISSUE #54549: middleware.ts treated as a client side component
### Verify canary release

- [X] I verified that the issue exists in the latest Next.js canary release

### Provide environment information

```bash
Operating System:
      Platform: darwin
      Arch: arm64
      Version: Darwin Kernel Version 22.5.0: Thu Jun  8 22:22:20 PDT 2023; root:xnu-8796.121.3~7/RELEASE_ARM64_T6000
    Binaries:
      Node: 16.20.0
      npm: 8.19.4
      Yarn: 1.22.19
      pnpm: 8.4.0
    Relevant Packages:
      next: 13.4.20-canary.6
      eslint-config-next: 13.4.19
      react: 18.2.0
      react-dom: 18.2.0
      typescript: 5.2.2
    Next.js Config:
      output: N/A
```

### Which area(s) of Next.js are affected? (leave empty if unsure)

Middleware / Edge (API routes, runtime)

### Link to the code that reproduces this issue or a replay of the bug

https://github.com/mansdahlstrom1/middleware-client-component-bug

### To Reproduce

1. Create new Next app
2. Add middleware 
3. mark middleware as `server-only` with `import "server-only";`

### Describe the Bug

When using middleware we expect the middleware to run server-side only and to be able to use features like [Cookies](https://nextjs.org/docs/app/api-reference/functions/cookies) etc. But we are not able to do this as Next seem to think that the middleware is running client side. 

### Expected Behavior

MIddleware can be marked as server-only. Server side only features work in middleware.ts file. 

OR

Clearer error message that explains the real error. 

### Which browser are you using? (if relevant)

116.0.5845.110

### How are you deploying your application? (if relevant)

_No response_

<sub>[NEXT-1607](https://linear.app/vercel/issue/NEXT-1607/middlewarets-treated-as-a-client-side-component)</sub>

### PR DESCRIPTION
**Pull request by huozhi**
Users want to use `server-only` to restrict the middleware / app routes / pages api, but now it's failing as we're treating them as different webpack layers, but validating the `server-only` only with server components layers.

Here we modify the rules a bit to let everyone can use "server-only" for the bundles that targeting server-side.

For next-swc transformer, we introduce the new option `bundleType` which only has `"server" | "client" | "default"` 3 values:
* - `server`  for server-side targets, like server components, app routes, pages api, middleware
* - `client`   for client components targets such as client components app pages, or page routes under pages directory.
* - `default` for environment like jest, we don't validate module graph with swc, replaced the `disable_checks` introduced  [#54891](https://github.com/vercel/next.js/pull/54891).

Refactor a bit webpack-config to adapt to the new rules, after that `server-only` will be able to used in the server-side targets conventions like middleware and `pages/api`

Fixes #43700
Fixes #54549
Fixes #52833

Closes NEXT-1616
Closes NEXT-1607
Closes NEXT-1385

### CODE REVIEW

**Comment by ijjk** [comment]
## Failing test suites 
Commit: 71d40c1189a2d50e8fff78780498c5d7f6c5e2fa

`pnpm test-dev test/development/acceptance-app/invalid-imports.test.ts` 
- Error Overlay invalid imports > should show error when using styled-jsx in server component
- Error Overlay invalid imports > should show error when external package imports client-only in server component
- Error Overlay invalid imports > should show error when external package imports server-only in client component

<details>
<summary>Expand output</summary>

  ● Error Overlay invalid imports › should show error when using styled-jsx in server component

    expect(received).toMatchInlineSnapshot(snapshot)

    Snapshot name: `Error Overlay invalid imports should show error when using styled-jsx in server component 1`

    Snapshot: "./app/comp2.js
    'client-only' cannot be imported from a Server Component module. It should only be used from a Client Component.

    The error was caused by using 'styled-jsx' in './app/comp2.js'. It only works in a Client Component but none of its parents are marked with \"use client\", so they're Server Components by default.

    Import trace for requested module:
    ./app/comp2.js
    ./app/comp1.js
    ./app/page.js"
    Received: null

      67 |
      68 |     expect(await session.hasRedbox(true)).toBe(true)
    > 69 |     expect(await session.getRedboxSource()).toMatchInlineSnapshot(`
         |                                             ^
      70 |         "./app/comp2.js
      71 |         'client-only' cannot be imported from a Server Component module. It should only be used from a Client Component.
      72 |

      at Object.<anonymous> (development/acceptance-app/invalid-imports.test.ts:69:45)

  ● Error Overlay invalid imports › should show error when external package imports client-only in server component

    expect(received).toMatchInlineSnapshot(snapshot)

    Snapshot name: `Error Overlay invalid imports should show error when external package imports client-only in server component 1`

    Snapshot: "./app/comp2.js
    'client-only' cannot be imported from a Server Component module. It should only be used from a Client Component.

    The error was caused by importing 'client-only-package/index.js' in './app/comp2.js'.

    Import trace for requested module:
    ./app/comp2.js
    ./app/comp1.js
    ./app/page.js"
    Received: null

      142 |
      143 |     expect(await session.hasRedbox(true)).toBe(true)
    > 144 |     expect(await session.getRedboxSource()).toMatchInlineSnapshot(`
          |                                             ^
      145 |         "./app/comp2.js
      146 |         'client-only' cannot be imported from a Server Component module. It should only be used from a Client Component.
      147 |

      at Object.<anonymous> (development/acceptance-app/invalid-imports.test.ts:144:45)

  ● Error Overlay invalid imports › should show error when external package imports server-only in client component

    expect(received).toMatchInlineSnapshot(snapshot)

    Snapshot name: `Error Overlay invalid imports should show error when external package imports server-only in client component 1`

    Snapshot: "./app/comp2.js
    'server-only' cannot be imported from a Client Component module. It should only be used from a Server Component.

    The error was caused by importing 'server-only-package/index.js' in './app/comp2.js'.

    Import trace for requested module:
    ./app/comp2.js
    ./app/comp1.js
    ./app/page.js"
    Received: null

      215 |
      216 |     expect(await session.hasRedbox(true)).toBe(true)
    > 217 |     expect(await session.getRedboxSource()).toMatchInlineSnapshot(`
          |                                             ^
      218 |         "./app/comp2.js
      219 |         'server-only' cannot be imported from a Client Component module. It should only be used from a Server Component.
      220 |

      at Object.<anonymous> (development/acceptance-app/invalid-imports.test.ts:217:45)

</details>


Read more about building and testing Next.js in [contributing.md](https://github.com/vercel/next.js/blob/canary/contributing.md).

File: packages/next/src/build/webpack-config.ts [hunk 1]
```diff
@@ -932,29 +917,62 @@ export default async function getBaseWebpackConfig(
...
       // This will cause some performance overhead but
       // acceptable as Babel will not be recommended.
-      [getSwcLoader({ hasServerComponents: false }), getBabelLoader()]
+      [
+        getSwcLoader({ hasServerComponents: false, bundleTarget: 'server' }),
+        getBabelLoader(),
+      ]
+
+  // client components layers: SSR + browser
+  const swcLoaderForClientLayers = [
```
Comment (ztanner) [inline packages/next/src/build/webpack-config.ts:943]: nit for consistency

```suggestion
  const swcLoaderForClientLayer = [
```

File: packages/next/src/build/webpack-config.ts [hunk 2]
```diff
@@ -2234,28 +2309,12 @@ export default async function getBaseWebpackConfig(
...
-                        // in the client layer.
-                        loader: 'next-flight-client-module-loader',
-                      },
-                      ...swcLoaderForClientLayer,
+                    issuerLayer: [
+                      WEBPACK_LAYERS.appPagesBrowser,
+                      WEBPACK_LAYERS.serverSideRendering,
                     ],
+                    exclude: [codeCondition.exclude],
+                    use: swcLoaderForClientLayers,
```
Comment (ztanner) [inline packages/next/src/build/webpack-config.ts:2317]:
```suggestion
                    use: swcLoaderForClientLayer,
```

File: packages/next/src/build/webpack-config.ts [hunk 3]
```diff
@@ -2024,6 +2041,66 @@ export default async function getBaseWebpackConfig(
...
+          },
+        },
+        {
+          issuerLayer: {
+            not: WEBPACK_LAYERS.GROUP.serverTarget,
+          },
+          resolve: {
+            // Error on server-only but allow client-only
+            alias: {
+              'server-only$': 'next/dist/compiled/server-only/index',
```
Comment (ztanner) [inline packages/next/src/build/webpack-config.ts:2066]: should this be pointing to `error` based on the comment above?

Comment (huozhi) [reply packages/next/src/build/webpack-config.ts:2066]: Yeah this exports files was bit different between client-only and server-only, sometimes confusing

**Review by shuding — APPROVED** [review]
Thanks for the code refactoring!

File: packages/next-swc/crates/core/tests/full.rs [hunk 4]
```diff
@@ -81,7 +81,7 @@ fn test(input: &Path, minify: bool) {
                 auto_modularize_imports: None,
                 optimize_barrel_exports: None,
                 optimize_server_react: None,
-                disable_checks: false,
+                bundle_target: String::from("default").into(),
```
Comment (sokra) [inline packages/next-swc/crates/core/tests/full.rs:84]: This should be an enum, shouldn't it?

Comment (huozhi) [reply packages/next-swc/crates/core/tests/full.rs:84]: yes, enum will be better, add a reminder for myself to refactor it later. thanks for suggestion 👍

**Comment by barryengineerapart** [comment]
@huozhi @timneutkens Just a big thank you from myself... This unblocks us for Next and so we can proceed. In general, how long does it take for canary to be cut into a new version.

However, thank you very much everyone. This has been tested and it works.

**Comment by huozhi** [comment]
@barryengineerapart glad to hear it's confirmed working! thanks

**Pull request merged**